services/elasticache/pkg/resource/replication_group/custom_update_api.go

```diff
@@ -490,6 +555,193 @@ func (rm *resourceManager) newUpdateShardConfigurationRequestPayload(
+	}
+	if resp.CacheClusters == nil {
+		return nil, nil
+	}
+
+	for _, cc := range resp.CacheClusters {
+		if cc == nil {
+			continue
+		}
+		return cc, nil
+	}
+	return nil, nil
+}
+
+// securityGroupIdsDiffer return true if
+// Security Group Ids differ between desired spec and latest (from cache cluster) status
+func (rm *resourceManager) securityGroupIdsDiffer(
+	desired *resource,
+	latest *resource,
+	latestCacheCluster *svcsdk.CacheCluster,
+) bool {
+	if desired.ko.Spec.SecurityGroupIDs == nil {
+		return false
+	}
+
+	desiredIds := []*string{}
+	for _, id := range desired.ko.Spec.SecurityGroupIDs {
+		if id == nil {
+			continue
+		}
+		var value string
+		value = *id
+		desiredIds = append(desiredIds, &value)
+	}
+	sort.Slice(desiredIds, func(i, j int) bool {
+		return *desiredIds[i] < *desiredIds[j]
+	})
+
+	latestIds := []*string{}
+	if latestCacheCluster.SecurityGroups != nil {
+		for _, latestSG := range latestCacheCluster.SecurityGroups {
+			if latestSG == nil {
+				continue
+			}
+			var value string
+			value = *latestSG.SecurityGroupId
+			latestIds = append(latestIds, &value)
+		}
+	}
+	sort.Slice(latestIds, func(i, j int) bool {
+		return *latestIds[i] < *latestIds[j]
+	})
+
+	if len(desiredIds) != len(latestIds) {
+		return true // differ
+	}
+	for index, desiredId := range desiredIds {
+		if *desiredId != *latestIds[index] {
+			return true // differ
+		}
+	}
+	// no difference
+	return false
+}
+
+// updateEngineVersion modifies the Replication Group to update securoty group Ids
+// it invokes ModifyReplicationGroup API to do so.
+func (rm *resourceManager) updateSecurityGroupIds(
+	ctx context.Context,
+	desired *resource,
+	latest *resource,
+) (*resource, error) {
+	input := &svcsdk.ModifyReplicationGroupInput{}
+
+	input.SetApplyImmediately(true)
+	if desired.ko.Spec.ReplicationGroupID != nil {
+		input.SetReplicationGroupId(*desired.ko.Spec.ReplicationGroupID)
+	}
+
+	if desired.ko.Spec.SecurityGroupIDs != nil {
+		ids := []*string{}
+		for _, id := range desired.ko.Spec.SecurityGroupIDs {
+			var value string
+			value = *id
+			ids = append(ids, &value)
+		}
+		input.SetSecurityGroupIds(ids)
+	}
+
+	resp, respErr := rm.sdkapi.ModifyReplicationGroupWithContext(ctx, input)
+	rm.metrics.RecordAPICall("UPDATE", "ModifyReplicationGroup-SecurityGroupsIds", respErr)
+	if respErr != nil {
+		rm.log.Error(respErr, "Error during ModifyReplicationGroup-SecurityGroupsIds")
+		return nil, respErr
+	}
+
+	return rm.provideUpdatedResource(desired, resp.ReplicationGroup)
+}
+
+// securityGroupIdsDiffer return true if
+// Engine Version differs between desired spec and latest (from cache cluster) status
+func (rm *resourceManager) engineVersionDiffer(
+	desired *resource,
+	latest *resource,
+	latestCacheCluster *svcsdk.CacheCluster,
+) bool {
+	if desired.ko.Spec.EngineVersion == nil {
+		return false
+	}
+	desiredEV := *desired.ko.Spec.EngineVersion
+
+	var latestEV string = ""
+	if latestCacheCluster.EngineVersion != nil {
```

Comment:
needs a nil-guard since `latestCacheCluster` can be `nil`

Reply:
done